for use in peergos

kevodwyer opened the pull request:
This includes a lot of fixes for issues discovered whilst trying to use nabu in Peergos.

ianopolous committed:
Make sure http proxy releases buffers

ianopolous committed:
Reduce test logspam

kevodwyer committed:
updates to support peergos

ianopolous commented on src/main/java/org/peergos/config/Config.java:
```diff
@@ -68,18 +78,10 @@ public Config defaultConfig(Optional<Supplier<Mount>> dataStoreSupplier) {
         MultiAddress apiAddress = new MultiAddress("/ip4/127.0.0.1/tcp/5001");
         MultiAddress gatewayAddress = new MultiAddress("/ip4/127.0.0.1/tcp/8080");
-        Optional<MultiAddress> proxyTargetAddress = Optional.of(new MultiAddress("/ip4/127.0.0.1/tcp/8000"));
+        Optional<MultiAddress> proxyTargetAddress = Optional.of(new MultiAddress("/ip4/127.0.0.1/tcp/8003"));
```
what's on 8003?

ianopolous commented on src/main/java/org/peergos/HttpProxyService.java:
```diff
@@ -48,7 +48,10 @@ public ProxyResponse proxyRequest(Multihash targetNodeId, ProxyRequest request)
+        Multiaddr finalAddress = targetAddressesOpt.get().toString().contains("/ipfs/") ? targetAddressesOpt.get()
+                : new Multiaddr(targetAddressesOpt.get() + "/ipfs/" + targetNodeId);
+
+        HttpProtocol.HttpController proxier = p2pHttpBinding.dial(node, finalAddress).getController().join();
```
We should be dialing all the addresses here, not arbitrarily the first one.

ianopolous commented on src/main/java/org/peergos/net/HttpProxyHandler.java:
```diff
@@ -12,20 +12,29 @@
 
     private static final String HTTP_REQUEST = "/http/";
+    private static final Function<String, Multihash> defaultPeerIdToMultiHashConverter = (peerIdStr) ->
+         Multihash.deserialize(PeerId.fromBase58(peerIdStr).getBytes());
```
Any reason not to just make this a static function? Can also just directly decode the base58, don't need to go via PeerId.

kevodwyer committed:
dial all addresses

kevodwyer committed:
change encoding of PeerId

ianopolous committed:
Dial all addrs in kademlia.

We could reach a node on the public dht via a local network address.

kevodwyer committed:
make access to router synchronized

kevodwyer committed:
set response headers

kevodwyer committed:
call flush on response

kevodwyer committed:
extract byte[] from request

ianopolous committed:
Fallback to DHT in block.get if no peers are provided

ianopolous committed:
Make getProviders also query the local provider store

ianopolous committed:
Add Blockstore.hasAny(raw multihash)

this is needed by the dht to answer if we are providing a block

ianopolous committed:
Fix adding ourselves to dht.getProviders

ianopolous committed:
Remove possibility for auth in want being Optional.of("")

ianopolous committed:
Send want blocks immediately in bitswap if the audience is small

ianopolous committed:
Make bitswap return auth with blocks

ianopolous committed:
Encode hex auth as bytes in bitswap

ianopolous committed:
Remove log spam from kademlia

ianopolous committed:
Fix auth in bitswap sent blocks

ianopolous committed:
Rename auth parameter in Want to authHex

ianopolous committed:
Remove log spam from connect exceptions during bootstrap

ianopolous committed:
Remove log spam from failed provides

ianopolous committed:
Remove bitswap log spam

ianopolous committed:
Handle error returning error

ianopolous committed:
Remove more log spam

ianopolous committed:
Remove more log spam

ianopolous committed:
Only update address book from identify results.

And do an identify on all new connections.

ianopolous committed:
Fix identify call on new connections

Make sure to add outbound connections to address book in kademlia so we have bootstrap nodes in there

Switch to TLS before noise

ianopolous committed:
Add the Identify protocol after starting the Host, so we know the concrete ports it is listening on.

This needs an upstream fix to make Host.addProtocolHandler work

ianopolous committed:
Include upstream fix to make Host.addProtocolHandler work

ianopolous committed:
Explicitly add supplied bootstrap addresses to address book

Adding the result of connecting to them is incorrect (wrong port)

ianopolous committed:
Add Identify to tests that need it

ianopolous committed:
Add Identify to more tests that need it

ianopolous committed:
Revert to noise over tls

ianopolous committed:
Add identify to bootstrap test

ianopolous committed:
Add peers from all connections to addressbook

ianopolous committed:
Remove RequestSender and use HttpProtocol.proxyRequest

ianopolous committed:
Use 1 HTTP reauest per stream in p2p http proxy. This removes head of line blocking issues.

Only set Host header in sender of p2p http requests, not in handler

ianopolous committed:
Fix test that was sending multiple http requests down same stream.
